provd/cmd/provd/daemon/config.go

```diff
@@ -0,0 +1,73 @@
+package daemon
+
+import (
+	"errors"
+	"fmt"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"github.com/ubuntu/decorate"
+)
+
+// initViperConfig adds config env variables and file support based on name prefix.
+func initViperConfig(name string, cmd *cobra.Command, vip *viper.Viper) (err error) {
+	defer decorate.OnError(&err, "can't load configuration")
+
+	// Handle configuration.
+	if v, err := cmd.Flags().GetString("config"); err == nil && v != "" {
+		vip.SetConfigFile(v)
+	} else {
+		vip.SetConfigName(name)
+		vip.AddConfigPath("./")
+		vip.AddConfigPath("$HOME/")
+		vip.AddConfigPath("/etc/provd/")
+		// Add the executable path to the config search path.
+		if binPath, err := os.Executable(); err != nil {
+			slog.Warn("Failed to get current executable path, not adding it as a config dir", "errorDetail", err)
```

Comment:
This is not the way to structure logs in Go (but I know the logging is really recent and you won’t find a lot of doc on it so far).

So, think of the first main message being really what you want to present to the user. I think here we are hiding in most of the cases the real underlying error. There is the risk of stuttering (and this is where a detail log/error piling up analyze is needed, and I must admit we didn’t do our clear diligence in all our projects yet).

But basically, think of the first message as "what do I want to display to the user". You will generally have to `fmt.Sprintf()` then to embed the underlying error.

Then, you have pair of tags for structured logging (mostly use in backends for searching/tagging logs). From some research from Google when building slog, most of projects have 1 tag in reality (the ID request), and up to 5 in very complex projects. Those are search fields for tracing, triaging, categorizing the type of errors, not details on the error itself.

So, in this case, we want to show the real error, on line 44, we want to display which configuration file is used in an human readable way (without tagging) and build that as a real english sentence. I hope that makes sense :)

Reply:
Ah I see! Yes, that is an abuse of tags on my part :smile:  So on line 44 we would want something like:
`slog.Info(fmt.Sprintf("Using configuration file: %v", vip.ConfigFileUsed()))`

In terms of stuttered logging, should I go through our info and warns and make sure we aren't duplicating information? So in this example, we already indicate on line 44 what config file we are using, which would be sufficient for a user to realize we couldn't find the one in the executable path?

Reply:
> Ah I see! Yes, that is an abuse of tags on my part 😄 So on line 44 we would want something like:
slog.Info(fmt.Sprintf("Using configuration file: %v", vip.ConfigFileUsed())) ?

Exactly! (I know the additional `fmt.Sprintf()` is annoying, but I’m sure our wrapper package taking a context and logger will help with this, once we get a chance to redo it.

> In terms of stuttered logging, should I go through our info and warns and make sure we aren't duplicating information? So in this example, we already indicate on line 44 what config file we are using, which would be sufficient for a user to realize we couldn't find the one in the executable path?
In case of error, I think it’s fine to tell what the issue is and repeat the config. The stutter we should aim at avoiding is:

`WARNING: could not get valid configuration file content for 'foo.bar': can't access property 'bla': error opening file 'foo:bar': could not stat foo.bar`

With error wrapping, this is really easy to get and to get sloppy with. More something to keep in mind for now when wrapping errors and then showing them as warning/error logs (I don’t think we have that issue in practice with info/debug).